### PR TITLE
improve geometry image kernels to support 2d mask inputs

### PR DESCRIPTION
**Pull request by pmeier**
In #6574, we realized that we can have 2d segmentation masks, i.e. `(H, W)`, and thus we cannot simply call the image kernel in case that accesses the number of channels, i.e. `C` in `C, H, W`. As a workaround we added (un-)squeezing logic to the mask kernels, e.g.

https://github.com/pytorch/vision/blob/4af683107f967d4d435be4180020989dd8c3019c/torchvision/transforms/v2/functional/_geometry.py#L307-L319

Since our kernels support arbitrary batch sizes, but the PyTorch kernels we call internally often require 4D inputs, we have additional reshaping logic inside the image kernels, e.g.

https://github.com/pytorch/vision/blob/4af683107f967d4d435be4180020989dd8c3019c/torchvision/transforms/v2/functional/_geometry.py#L221

However, this alone is not sufficient. 3d detection masks, i.e. `(N, H, W)`, can have `N==0`. This might for example happen after `SanitizeBoundingBoxes`. For that case our image kernels have additional checks, e.g.

https://github.com/pytorch/vision/blob/4af683107f967d4d435be4180020989dd8c3019c/torchvision/transforms/v2/functional/_geometry.py#L220

This PR simplifies this by merging all this logic inside the image kernel.

cc @vfdev-5

### CODE REVIEW

File: torchvision/transforms/v2/functional/_geometry.py [hunk 1]
```diff
@@ -210,64 +210,73 @@ def resize_image(
...
+    if image.ndim > 2:
+        num_channels = image.shape[-3]
+        output_shape = image.shape[:-3] + (num_channels, new_height, new_width)
+    else:
+        num_channels = 1
+        output_shape = [new_height, new_width]
+
+    numel = image.numel()
+    if numel == 0:
+        return image.reshape(output_shape)
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:228]: This is the only new functionality. Since we now compute the `output_shape` upfront, we can make the `numel == 0` case a shortcut and return early. Due to this, I removed the `if / else` branching below and thus could remove one level of indent. Unfortunately, `git` is not picking up on that and thus this diff looks larger than it is. There is no functional change to the core of `F.resize_image`.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 2]
```diff
@@ -554,19 +552,36 @@ def _compute_affine_output_size(matrix: List[float], w: int, h: int) -> Tuple[in
 
 
 def _apply_grid_transform(img: torch.Tensor, grid: torch.Tensor, mode: str, fill: _FillTypeJIT) -> torch.Tensor:
+    input_shape = img.shape
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:555]: Putting the handling avoids repeating this in every kernel that uses this as core action, i.e. `F.affine`, `F.rotate`, `F.perspective`, and `F.elastic`.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 3]
```diff
@@ -625,10 +640,10 @@ def _assert_grid_transform_inputs(
...
 def _affine_grid(
     theta: torch.Tensor,
-    w: int,
-    h: int,
-    ow: int,
-    oh: int,
+    input_width: int,
+    input_height: int,
+    output_width: int,
+    output_height: int,
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:646]: Driveby to make this easier to understand.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 4]
```diff
@@ -1184,14 +1160,23 @@ def _pad_with_scalar_fill(
     fill: Union[int, float],
     padding_mode: str,
 ) -> torch.Tensor:
-    shape = image.shape
-    num_channels, height, width = shape[-3:]
+    old_height, old_width = image.shape[-2:]
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:1163]: I've put the handling only in `_pad_with_scalar_fill` since that is the only thing that we can hit for masks:

https://github.com/pytorch/vision/blob/4af683107f967d4d435be4180020989dd8c3019c/torchvision/transforms/v2/functional/_geometry.py#L1268-L1269

However, we could also move it up to `pad_image`. LMK if you prefer that.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 5]
```diff
@@ -1401,18 +1373,7 @@ def _crop_bounding_boxes_dispatch(
...
-    else:
-        needs_squeeze = False
-
-    output = crop_image(mask, top, left, height, width)
-
-    if needs_squeeze:
-        output = output.squeeze(0)
-
-    return output
+    return crop_image(mask, top, left, height, width)
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:1376]: Since the introduction of the (un-)squeeze logic here, we seem to have refactored the image kernel. I was able to remove that without touching the image kernel.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 6]
```diff
@@ -1452,11 +1413,11 @@ def perspective(
     )
 
 
-def _perspective_grid(coeffs: List[float], ow: int, oh: int, dtype: torch.dtype, device: torch.device) -> torch.Tensor:
-    # https://github.com/python-pillow/Pillow/blob/4634eafe3c695a014267eefdce830b4a825beed7/
-    # src/libImaging/Geometry.c#L394
+def _perspective_grid(
+    coeffs: List[float], width: int, height: int, dtype: torch.dtype, device: torch.device
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:1417]: Driveby to make this easier readable.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 7]
```diff
@@ -2039,18 +1935,7 @@ def _center_crop_bounding_boxes_dispatch(
 
 @_register_kernel_internal(center_crop, tv_tensors.Mask)
 def center_crop_mask(mask: torch.Tensor, output_size: List[int]) -> torch.Tensor:
-    if mask.ndim < 3:
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:2042]: Same as for `crop_mask`: (un-)squeezing logic was no longer needed.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 8]
```diff
@@ -210,64 +210,73 @@ def resize_image(
...
-        dtype = image.dtype
-        acceptable_dtypes = [torch.float32, torch.float64]
-        if interpolation == InterpolationMode.NEAREST or interpolation == InterpolationMode.NEAREST_EXACT:
-            # uint8 dtype can be included for cpu and cuda input if nearest mode
+    if image.ndim > 2:
+        num_channels = image.shape[-3]
+        output_shape = image.shape[:-3] + (num_channels, new_height, new_width)
+    else:
+        num_channels = 1
+        output_shape = [new_height, new_width]  # type: ignore[assignment]
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_geometry.py:224]: Need to silence `mypy` here, because torchscript is doing something stupid here.

```python
image.shape[:-3] + (num_channels, new_height, new_width)
```

is a tuple, but internally this is represented as list. Meaning, I need to match that in the `else` branch as well. But `mypy` is picking up on that and complains that we now have a list or a tuple in eager mode depending on the branch.

There are two more patterns of this.

**Review by vfdev-5 — APPROVED** [review]
Looks ok to me, thanks @pmeier

**Review by NicolasHug — COMMENTED** [review]
I fail to see the benefit of this PR.

> This PR simplifies this

It's really hard to assess that from the diff. I guess it's because the `diff` tool struggles to make a clean diff and it's difficult to figure what got shuffled around. But the drive-bys don't help either.

Before spending more time on a review, are we absolutely sure we don't take a perf hit on the image kernels?

**Comment by NicolasHug** [comment]
Trying to summarize my understanding.

This PR transforms this boilerplate but simple logic, which is repeated across mask kernels:


https://github.com/pytorch/vision/blob/4af683107f967d4d435be4180020989dd8c3019c/torchvision/transforms/v2/functional/_geometry.py#L307-L319


into:

- mask-specific logics **within the image kernels**, therefore turning the specialized image kernels into more general-purpose kernels
- logics that are in different places but more importantly logics that are **different for each kernel**, because they have to adapt to of the specific image kernels
- this also adds support for 2D images in the image kernels. Not only this is something we don't need, but I'm also not sure whether this is something we actually want: we'd have to keep supporting it in the future
